### PR TITLE
environment views: dependents before dependencies, resolve identical file conflicts

### LINKED ISSUE #12482: File conflicts can occur in views if the conflicting packages do not cause concretization errors
I've noticed some unexpected behavior if an environment (or rather its view) contains two packages that have file conflicts, but do not conflict themselves. More specifically, `metis` and `scotch^+metis` both provide `include/metis.h` but do not conflict otherwise - it's perfectly possible  to add both to an environment and install it, without Spack complaining at all.

However, the files are _not_ directly compatible and while METIS' `metis.h` should be useable with `scotchmetis` (that's the recommended way by Scotch), the converse is not true and SCOTCH's `metis.h` does not contain all functions METIS offers, rendering METIS impaired.

Additionally, if you remove `scotch` from the environment afterwards, the conflicting files will be deleted and any install command afterwards will error like this:
```
===> Error: /mypathtospack/var/spack/environments/myenv/.spack-env/include/metis.h is not a link tree!
```
This seems to be a result of spack assuming that these files should exist if `metis` were installed properly. The situation is identical between `parmetis` and `spack` with their respective overlap.

Repro:
```
spack env create myenv
spack env activate myenv
spack add metis parmetis scotch
spack install
```

### PR DESCRIPTION
**Pull request by haampie**
This PR resolves two current issues with environment views:

1. Ensure correct linking order: dependents before dependencies
2. Resolve a class of file-file conflicts, that causes self-referential
  symlinks to be generated for copy type views.

This is useful in general, but certainly needed in #40773, where the package
`python-venv` contains a "conflicting" symlink

```
<python-venv>/bin/python3 -> <python>/bin/python3
```

(It's not very useful yet for ongoing work related to `gcc-runtime`, where
multiple, ABI-compatible `gcc-runtime` nodes are going to be allowed in the
same dag. There, the newest `gcc-runtime` should be linked, topological
order does not help there, unless we make them each other's link deps...)

#### Ordering

This PR ensures that when a dependent and a dependency have *different*
executables at `<prefix>/bin/exe`, the dependent's executable is linked
into the view. This coincides with how `PATH` entries are ordered on
`spack load` or `spack env activate` without views.

#### File-file conflicts for identical files

It can also be that a dependent and a dependency have *identical*
executables at `<prefix>/bin/exe`, in the sense that `os.samefile(f, g)` is true.
That happens when their `stat` following symlinks have identical
`(st_ino, st_dev)` -- it does not involve expensive content comparison.
Or in less technical terms: the dependent hardlinks or symlinks a file of its
dependency at the same relative path.

That used to result in a file-file conflict, but this PR resolves the conflict by
only registering the file from the dependency to be linked.

Doing that solves a bug in copy type views. Previously we would "copy" the
symlink and relocate its target path. In the above case that results in a
self-referential symlink in the view. With this PR, the underlying file is taken,
and the actual file is copied, as one would expect.

The strategy applies also to symlink type views, where nothing really changes,
except that the symlink chain would be one link shorter.

### CODE REVIEW

File: lib/spack/spack/filesystem_view.py [hunk 1]
```diff
@@ -666,9 +667,6 @@ def add_specs(self, *specs, **kwargs):
                 tty.warn("Skipping external package: " + s.short_spec)
         specs = [s for s in specs if not s.external]
 
-        if kwargs.get("exclude", None):
-            specs = set(filter_exclude(specs, kwargs["exclude"]))
-
```
Comment (scheibelp) [inline lib/spack/spack/filesystem_view.py]: Are you removing this because nothing ever calls `add_specs` with this kwarg (i.e. this change is unrelated to this PR)? I see a use of this in `lib/spack/spack/cmd/view.py`

File: lib/spack/spack/environment/environment.py [hunk 2]
```diff
@@ -715,7 +715,7 @@ def regenerate(self, concretized_root_specs):
         # Create a new view
         try:
             fs.mkdirp(new_root)
-            view.add_specs(*specs, with_dependencies=False)
+            view.add_specs(*specs)
```
Comment (scheibelp) [inline lib/spack/spack/environment/environment.py]: Why remove this? `specs_for_view` grabs all the dependencies, so I think this arg was to avoid that. Actually there is a slight behavioral change as `get_dependencies` will add all dependencies regardless of deptype.

File: lib/spack/llnl/util/link_tree.py [hunk 3]
```diff
@@ -150,15 +151,27 @@ def visit_file(self, root: str, rel_path: str, depth: int) -> None:
...
-                    src_a=os.path.join(src_a_root, src_a_relpath),
-                    src_b=os.path.join(root, rel_path),
+
+            src_a = os.path.join(src_a_root, src_a_relpath)
+            src_b = os.path.join(root, rel_path)
+
+            if os.path.samefile(src_a, src_b):
+                # Delete and re-insert to retain correct self.files ordering.
+                del self.files[proj_rel_path]
+                self.files[proj_rel_path] = (root, rel_path)
```
Comment (scheibelp) [inline lib/spack/llnl/util/link_tree.py]: I'm finding this difficult to understand. Reading it out: "src_a and src_b are the same file, so use src_b"... (In my head: "how can this matter?") - I think this works for https://github.com/spack/spack/pull/40773#issuecomment-1914990486 because 

* we traverse from parent to child
* we visit python second (after `python-venv`), which is a file vs. a symlink
* we overwrite with that (our `src_b`)
* when it comes time to `view_copy`, it's not a symlink to a relative `.../bin/python` so it doesn't relocate the link

So reversing the order of traversals would recreate https://github.com/spack/spack/pull/40773#issuecomment-1914990486. However unlikely, it seems we might as well make this robust to that.

I'm inclined to say this should explicitly resolve the realpath of either, then make *that* the new `self.files[proj_rel_path]` (and if they are both not symlinks, there should be no need to overwrite, although I think that cannot happen). 

Other option: pick whichever one is a file, and use that

File: lib/spack/llnl/util/link_tree.py [hunk 4]
```diff
@@ -150,15 +151,27 @@ def visit_file(self, root: str, rel_path: str, depth: int) -> None:
...
-                MergeConflict(
-                    dst=proj_rel_path,
-                    src_a=os.path.join(src_a_root, src_a_relpath),
-                    src_b=os.path.join(root, rel_path),
+
+            src_a = os.path.join(src_a_root, src_a_relpath)
+            src_b = os.path.join(root, rel_path)
+
+            if os.path.samefile(src_a, src_b):
+                # Delete and re-insert to retain correct self.files ordering.
```
Comment (scheibelp) [inline lib/spack/llnl/util/link_tree.py]: From what I knew, `self.files` is a `{}`, which does not have ordering guarantees

Apparently, for Python >= 3.7, it is ordered. But we support older versions (https://spack.readthedocs.io/en/latest/getting_started.html#python says 3.2+ but our changelog and https://spack.readthedocs.io/en/latest/getting_started.html#system-prerequisites indicate 3.6+).

Either way, if we depend on ordering, we'll want an [OrderedDict](https://docs.python.org/3/library/collections.html#collections.OrderedDict)

File: lib/spack/spack/test/cmd/env.py [hunk 5]
```diff
@@ -3714,3 +3714,27 @@ def test_environment_created_from_lockfile_has_view(mock_packages, temporary_sto
     # Make sure the view was created
     with ev.Environment(env_b) as e:
         assert os.path.isdir(e.view_path_default)
+
+
+def test_env_view_resolves_identical_file_conflicts(tmp_path, install_mockery, mock_fetch):
```
Comment (scheibelp) [inline lib/spack/spack/test/cmd/env.py:3896]: (minor) suggested name: `test_view_relative_symlink_is_cyclical`

The logic should also resolves cases like `x->y->z`, and x/y both create `bin/w -> <z-prefix>/bin/w` (which is not cyclic) but this test is more specific

**Review by scheibelp — CHANGES_REQUESTED** [review]
In addition to the comments, I think `view_copy` should raise an error when creating self-referential symlinks

File: lib/spack/llnl/util/link_tree.py (hunk 3 above)
Comment (haampie) [reply lib/spack/llnl/util/link_tree.py]: > So reversing the order of traversals would recreate https://github.com/spack/spack/pull/40773#issuecomment-1914990486. However unlikely, it seems we might as well make this robust to that.

The `test_env_view_ignores_different_file_conflicts` test would fail, so it's robust already.

> I'm inclined to say this should explicitly resolve the realpath of either, then make that the new self.files[proj_rel_path] 

That would be completely wrong for multiple reasons, realpath would be a ... real mistake

1. `realpath(x)` may not exist, causing link failure later
2. It can cause redundant file copies in copy-type views where we *can* copy a symlink in a chain like `<prefix A>/x -> <prefix B>/x -> <prefix C>/y`.
3.  If `<prefix A>/x -> <prefix B>/x -> /system/x -> /system/y`, and `/system/x` changes, it would not be reflected in the view, breaking it. E.g. `curl -> openssl certs=system -> /etc/ssl/certs/*` or so. (For internal spack symlinks i'm assuming prefixes are immutable).
4. `src_root` has to be the prefix of the current Spack package (but that may be an implementation detail)

> Other option: pick whichever one is a file, and use that

First, that doesn't work when both are links. Second, realpath triggers O(depth + chain length) readlink calls which I want to avoid, mine is just two stat calls. This prefix scan is supposed to be fast.

So, let's stick to samefile.

[Another pitfall with realpath is you'd need to worry about case insensitive file systems, but to be fair that's another bug already present anyways, since normcase is missing when keying self.files]

File: lib/spack/spack/filesystem_view.py (hunk 1 above)
Comment (haampie) [reply lib/spack/spack/filesystem_view.py]: It's unused for years, i'm getting rid of `**kwargs` in the process, which has caused several bugs in the past due to typos. If you think that is unrelated I'm willing to do so elsewhere, but while adding type hints, mypy started complaining about the fact that `specs` is a union of list and spec, not a list, so I decided to resolve that by just deleting the unused code that triggers it.

File: lib/spack/spack/environment/environment.py (hunk 2 above)
Comment (haampie) [reply lib/spack/spack/environment/environment.py]: > slight behavioral change as get_dependencies will add all dependencies regardless of deptype.

If so then tests would have failed.

Traversing, selecting, excluding, and ordering the specs is done since about 5 years in ViewDescriptor where the info to do so is. `with_dependencies` does nothing. It's removed here because I dropped `**kwargs` per other comment.

The issue is that the `FilesystemView` interface is too big. There are two implementations, they are never used generically, not even in tests, and you can't swap in one for the other in the two use cases (spack view /spack env) because they're too different. The scope of this PR is not to do something about it.

File: lib/spack/llnl/util/link_tree.py (hunk 4 above)
Comment (haampie) [reply lib/spack/llnl/util/link_tree.py]: Thanks, I thought that was an issue of the past.

Edit: dicts are ordered in cpython 3.6, it's just that that was set in stone from 3.7 onwards. Do we care? (I asked on Slack, looks like "no")

File: lib/spack/spack/test/cmd/env.py (hunk 5 above)
Comment (haampie) [reply lib/spack/spack/test/cmd/env.py:3896]: `test_view_relative_symlink_is_cyclical` I don't really get the name. You mean `is_not_cyclical`?

OK, I think I also wanna ensure tests fail if the implementation was based on `realpath`, so:

```
<prefix a>/bin/
  x -> <prefix c>/bin/x # your suggestion
  y -> <prefix b>/bin/y


<prefix b>/bin/
  x -> <prefix c>/bin/x
  y -> <prefix c>/bin/x

<prefix c>/bin/
  x
```

and check that the `y` of `<prefix b>/bin/y` is linked. Test might be too strict  as integration test, cause it doesn't matter much what `y` is linked, but in case of future overhaul of views, I think it's fine to be strict.

File: lib/spack/llnl/util/link_tree.py (hunk 3 above)
Comment (scheibelp) [reply lib/spack/llnl/util/link_tree.py]: >> Other option: pick whichever one is a file, and use that

> First, that doesn't work when both are links. 

If they're both links, it should be fine to use either

> Second, realpath triggers O(depth + chain length) readlink calls

Could we `lstat` the files to determine whether they are links (which I think would involve fewer calls than `realpath`):

```
if samefile() and xor(islink(x) for x in src_a, src_b):
  # one of these is a symlink to the other, use the original file
```

?

As a general question: Doesn't samefile have to do the same thing as realpath? How could `samefile` work if it's not chasing down the link targets (transitively)?

Comment (haampie) [reply lib/spack/llnl/util/link_tree.py]: The visitor already has `visit_symlink`.

> If they're both links, it should be fine to use either

Then why suggest to distinguish instead of leaving it at `os.samefile`?

> Doesn't samefile have to do the same thing as realpath? How could samefile work if it's not chasing down the link targets (transitively)?

The kernel can avoid chasing through file system caches you can't have in userland, cause there you don't know how to invalidate them. But also it's about not paying syscall overhead, and slow python loops.

File: lib/spack/llnl/util/link_tree.py (hunk 4 above)
Comment (scheibelp) [reply lib/spack/llnl/util/link_tree.py]: `link_tree` has existed for a while and used `dict`, and didn't depend on ordering. So being careful here seems to be introducing new API guarantees. I'm inclined to say that:

* For now don't explicitly `del`, or
* Add a comment here about how we want to improve `link_tree`'s behavior to do an ordered traversal and we are leaning on the (implementation-specific) details of python dictionaries

File: lib/spack/llnl/util/link_tree.py (hunk 3 above)
Comment (scheibelp) [reply lib/spack/llnl/util/link_tree.py]: > The visitor already has visit_symlink.

In that case the check would be:

```
if samefile() and islink(src_a):
  # src_a is a symlink to src_b, which is a file: use the file.
  # We do this because src_a would become self-referential
  # if relocated relative to the link_tree's root
```

I think this comment would be useful compared to

```
            # Resolve file-file conflicts when two files are identical: this deals with cases of
            # of symlink to symlink, symlink to regular file and hardlinked files, where in
            # principle the file we pick is arbitrary. However, for environment views that copy,
            # we want the file to be copied, not the symlink, so we pick the underlying file, which
            # should be the current, assuming proper ordering.
```

(the former gets into the why, and avoids caring about the cases we don't need to change)

> Then why suggest to distinguish instead of leaving it at os.samefile

Doing something that doesn't need to be done (i.e. replacing with `src_b` when we don't need to) is confusing in this case: it obfuscates the reasons for doing it.

I'm open to other approaches to making this less confusing (e.g. the comment I suggested above would help).
Perhaps (also) putting this in the docstring for `SourceMergeVisitor` (and `LinkTree.merge`) would help: `test_source_merge_visitor_does_not_register_identical_file_conflicts` enforces that behavior for the tests, but the behavior of `merge` for these types of issues seems like something that ought to be explicit in the method's description.

File: lib/spack/llnl/util/link_tree.py (hunk 4 above)
Comment (haampie) [reply lib/spack/llnl/util/link_tree.py]: The order is relied on in `SimpleFilesystemView` for about two years. This PR does not introduce new behavior or API. The `del` is necessary, if you remove it a test fails.

If you are concerned about this detail, can't you deal with it in a separate PR? In principle this PR can be backported as a bugfix, I don't want to lay out a plan of future improvements in comments, cause I don't think it needs to be, I'd rather leave that to you since you seem concerned about it.

File: lib/spack/llnl/util/link_tree.py (hunk 3 above)
Comment (haampie) [reply lib/spack/llnl/util/link_tree.py:174]: To be honest I don't think `islink(src_a)` improves things _that_ much, cause you still need to explain the assumption on order: dependents before dependencies, symlinks only go from dependent -> dependency. 

The conflicting symlink to conflicting symlink to conflicting file case is just _so_ incredibly niche, I don't get why you care about "not doing something that doesn't need to be done". Also note that none of this is strictly necessary for copy type environments.

I'll try and write a concise and less technical comment based on your suggestion, that hopefully resolves the confusion.

File: lib/spack/llnl/util/link_tree.py (hunk 4 above)
Comment (scheibelp) [reply lib/spack/llnl/util/link_tree.py]: > The del is necessary, if you remove it a test fails.

Are you saying tests other than the ones added in this PR would fail if this `del`+reinsert logic was just an override?

Or that the logic in general would fail if this overwrite-without-delete occurred (because perhaps the dirs need to be made before the files in them)?

I'm not sure how either would be a problem: if it was ok to have `src_a` at this point, it seems like it's always OK to have `src_b` inserted at the same time.

Comment (haampie) [reply lib/spack/llnl/util/link_tree.py]: Files are later grouped by prefix. Group by assumes consecutive files from same prefix.

File: lib/spack/llnl/util/link_tree.py (hunk 3 above)
Comment (scheibelp) [reply lib/spack/llnl/util/link_tree.py]: > To be honest I don't think islink(src_a) improves things that much,

This is convoluted and I'm trying to find any way I can to make it less so. If you think this makes it even the slightest bit clearer, then it's worthwhile. Without checking `islink` the comment I suggested is imprecise and therefore less useful.

Zooming out, doing this in `add_files_to_view` would have been more clear because the concerns would have been addressed precisely where they arise. Another mechanism would be for example registering a conflict-resolution callback with the visitor that views could add their own implementation to. There's plenty of angles that avoid the requirement for discussing actions that only make sense for isolated users of this, so we could also consider those if it's impossible to make this particular approach more clear.

> cause you still need to explain the assumption on order: dependents before dependencies, symlinks only go from dependent -> dependency.

I thought the `islink(src_a)` part of `if samefile and islink(src_a)` would make it work in both directions (we only replace with `src_b` if `src_b` is the file.

File: lib/spack/llnl/util/link_tree.py (hunk 4 above)
Comment (scheibelp) [reply lib/spack/llnl/util/link_tree.py]: OK, how about we just augment the comment to point to this function (`_source_merge_visitor_to_merge_map`). Ideally we'd also update the docstring for SourceMergeVisitor to make this official. IMO this is also a good cautionary tale against reaching into the members of objects.

I want it to be clear we depend on preservation of order as an internal implementation detail. Ideally we could have a `_replace_file_item` function that codifies this internal constraint.

File: lib/spack/spack/test/cmd/env.py (hunk 5 above)
Comment (scheibelp) [reply lib/spack/spack/test/cmd/env.py:3896]: > I don't really get the name. You mean `is_not_cyclical`?

(FWIW feel free to reject/collapse/resolve this - it is minor)

I meant "would be cyclical except we avoid that and that's what this test is checking"

My thought was this is the test that deals with symlinks that would become self-referential. All this effort to fix the problem and I think the test name should more-clearly indicate this is the key behavior we want to check.

`resolves_identical_file_conflicts` doesn't evoke that for me

**Comment by haampie** [comment]
@scheibelp I think I've addressed your concerns in a reasonable way. Would appreciate it if you would get this PR to land.

File: lib/spack/llnl/util/link_tree.py (hunk 3 above)
Comment (haampie) [reply lib/spack/llnl/util/link_tree.py]: > would make it work in both directions 

True. I've done the reverse now since that information is directly available, and I can't care so much about the hardlink case (neither about the symlink case, but well): if current file clashes, but identical, and regular file, take it.

If you care about the hardlink case too, feel free to submit a follow-up pr.

File: lib/spack/llnl/util/link_tree.py [hunk 6]
```diff
@@ -134,38 +136,54 @@ def before_visit_symlinked_dir(self, root: str, rel_path: str, depth: int) -> bo
...
+                    MergeConflict(dst=proj_rel_path, src_a=src_a, src_b=src_b)
                 )
-            )
+                return
+
+            if not symlink:
+                # Remove the link in favor of the actual file. The del is necessary to maintain the
+                # order of the files dict, which is grouped by root.
+                del self.files[proj_rel_path]
+                self.files[proj_rel_path] = (root, rel_path)
```
Comment (scheibelp) [inline lib/spack/llnl/util/link_tree.py:178]: A comment like `# src_b in tuple form` would help

File: lib/spack/llnl/util/link_tree.py [hunk 7]
```diff
@@ -134,38 +136,54 @@ def before_visit_symlinked_dir(self, root: str, rel_path: str, depth: int) -> bo
...
+                # Distinct files produce a conflict.
+                self.file_conflicts.append(
+                    MergeConflict(dst=proj_rel_path, src_a=src_a, src_b=src_b)
                 )
-            )
+                return
+
+            if not symlink:
+                # Remove the link in favor of the actual file. The del is necessary to maintain the
+                # order of the files dict, which is grouped by root.
```
Comment (scheibelp) [inline lib/spack/llnl/util/link_tree.py:176]: Given that we are bound to internal implementation details IMO we might as well point at the exact culprit instead of hinting at it, which is why I suggested mentioning `_source_merge_visitor_to_merge_map` explicitly (a reader of this code may be curious about this, so they look for `groupby` in this file, maybe followed by `group_by` and maybe it occurs to them to look outside of the file, or maybe not).

File: lib/spack/llnl/util/link_tree.py [hunk 8]
```diff
@@ -134,38 +136,54 @@ def before_visit_symlinked_dir(self, root: str, rel_path: str, depth: int) -> bo
...
+            if not samefile:
+                # Distinct files produce a conflict.
+                self.file_conflicts.append(
+                    MergeConflict(dst=proj_rel_path, src_a=src_a, src_b=src_b)
                 )
-            )
+                return
+
+            if not symlink:
+                # Remove the link in favor of the actual file. The del is necessary to maintain the
```
Comment (scheibelp) [inline lib/spack/llnl/util/link_tree.py:175]: I would recommend modifying:

> Remove the link in favor of the actual file (src_b). The del is necessary to maintain the...

resolving what "the actual file" means requires reading from the beginning of the function.

File: var/spack/repos/builtin.mock/packages/view-ignore-conflict/package.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import os
+
+from spack.package import *
+
+
+class ViewIgnoreConflict(Package):
```
Comment (scheibelp) [inline var/spack/repos/builtin.mock/packages/view-ignore-conflict/package.py:11]: (minor) FWIW: I started finding it cumbersome to define multiple `builtin.mock` packages so I created a mechanism to define them as strings in whatever test file needs them. See for example https://github.com/spack/spack/pull/41406/files#diff-3d650e694e939950596b60fd4356346e0109f0d8171b92720a75e3664952d44e

**Review by scheibelp — APPROVED** [review]
Suggested commit message:

```
Views: dependents before dependencies, resolve identical file conflicts

Fix two separate problems:

1. We want to always visit parents before children while creating views
   (when it comes to ignoring conflicts, the first instance generated in
   the view is chosen, and we want the parent instance to have precedence).
   Our preorder traversal does not guarantee that, but our topological-
   order traversal does.
2. For copy style views with packages x depending on y, where
   <x-prefix>/foo is a symlink to <y-prefix>/foo, we want to guarantee
   that:
   * A conflict is not registered
   * <y-prefix>/foo is chosen (otherwise, the "foo" symlink would become
     self-referential if relocated relative to the view root)

   Note that
   * This is an exception to [1] (in this case the dependency instance
     overrides the dependent)
   * Prior to this change, if "foo" was ignored as a conflict, it was
     possible to create this self-referential symlink

Add tests for each of these cases
```

File: var/spack/repos/builtin.mock/packages/view-ignore-conflict/package.py (hunk 9 above)
Comment (haampie) [reply var/spack/repos/builtin.mock/packages/view-ignore-conflict/package.py:11]: I like that but would really like to just write

```python
def test_smth(create_package):
  class TestPackage(Package):
    def install(...):
      pass
  create_package(TestPackage)
```

which may or may not work with `inspect.getsource(TestPackage)`, but really would be great if supported w/o a package.py
